File: adapters/AppLovin/AppLovinAdapter/GADMAdapterAppLovinBannerDelegate.m
// Copyright 2019 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "GADMAdapterAppLovinBannerDelegate.h"
#import "GADMAdapterAppLovin.h"
#import "GADMAdapterAppLovinUtils.h"

@implementation GADMAdapterAppLovinBannerDelegate {
  /// AppLovin banner ad renderer to which the events are delegated.
  __weak GADMAdapterAppLovin *_parentAdapter;
}

#pragma mark - Initialization

- (nonnull instancetype)initWithParentAdapter:(nonnull GADMAdapterAppLovin *)parentAdapter {
  self = [super init];
  if (self) {
    _parentAdapter = parentAdapter;
  }
  return self;
}

#pragma mark - Ad Load Delegate

- (void)adService:(nonnull ALAdService *)adService didLoadAd:(nonnull ALAd *)ad {
  GADMAdapterAppLovin *parentAdapter = _parentAdapter;
  [GADMAdapterAppLovinUtils log:@"Banner did load ad: %@", ad];
  [parentAdapter.adView render:ad];
  [parentAdapter.connector adapter:parentAdapter didReceiveAdView:parentAdapter.adView];
}

- (void)adService:(nonnull ALAdService *)adService didFailToLoadAdWithError:(int)code {
  GADMAdapterAppLovin *parentAdapter = _parentAdapter;
  NSError *error = GADMAdapterAppLovinSDKErrorWithCode(code);
  [parentAdapter.connector adapter:parentAdapter didFailAd:error];
}

#pragma mark - Ad Display Delegate

- (void)ad:(nonnull ALAd *)ad wasDisplayedIn:(nonnull UIView *)view {
  [GADMAdapterAppLovinUtils log:@"Banner displayed"];
}

- (void)ad:(nonnull ALAd *)ad wasHiddenIn:(nonnull UIView *)view {
  [GADMAdapterAppLovinUtils log:@"Banner dismissed"];
}

- (void)ad:(nonnull ALAd *)ad wasClickedIn:(nonnull UIView *)view {
  [GADMAdapterAppLovinUtils log:@"Banner clicked"];
  GADMAdapterAppLovin *parentAdapter = _parentAdapter;
  [parentAdapter.connector adapterDidGetAdClick:parentAdapter];
}

#pragma mark - Ad View Event Delegate

- (void)ad:(nonnull ALAd *)ad didPresentFullscreenForAdView:(nonnull ALAdView *)adView {
  [GADMAdapterAppLovinUtils log:@"Banner presented fullscreen"];
  GADMAdapterAppLovin *parentAdapter = _parentAdapter;
  [parentAdapter.connector adapterWillPresentFullScreenModal:parentAdapter];
}

- (void)ad:(nonnull ALAd *)ad willDismissFullscreenForAdView:(nonnull ALAdView *)adView {
  [GADMAdapterAppLovinUtils log:@"Banner will dismiss fullscreen"];
  GADMAdapterAppLovin *parentAdapter = _parentAdapter;
  [parentAdapter.connector adapterWillDismissFullScreenModal:parentAdapter];
}

- (void)ad:(nonnull ALAd *)ad didDismissFullscreenForAdView:(nonnull ALAdView *)adView {
  [GADMAdapterAppLovinUtils log:@"Banner did dismiss fullscreen"];
  GADMAdapterAppLovin *parentAdapter = _parentAdapter;
  [parentAdapter.connector adapterDidDismissFullScreenModal:parentAdapter];
}

- (void)ad:(nonnull ALAd *)ad willLeaveApplicationForAdView:(nonnull ALAdView *)adView {
  [GADMAdapterAppLovinUtils log:@"Banner left application"];
  GADMAdapterAppLovin *parentAdapter = _parentAdapter;
  [parentAdapter.connector adapterWillLeaveApplication:parentAdapter];
}

- (void)ad:(nonnull ALAd *)ad
    didFailToDisplayInAdView:(nonnull ALAdView *)adView
                   withError:(ALAdViewDisplayErrorCode)code {
  [GADMAdapterAppLovinUtils log:@"Banner failed to display: %ld", (long)code];
}

@end

File: adapters/AppLovin/AppLovinAdapter/GADMAdapterAppLovinBannerDelegate.swift
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import AppLovinSDK
import Foundation
import GoogleMobileAds

@MainActor
@objc(GADMAdapterAppLovinBannerDelegate)
public class GADMAdapterAppLovinBannerDelegate: NSObject, @preconcurrency ALAdLoadDelegate,
  @preconcurrency ALAdDisplayDelegate, @preconcurrency ALAdViewEventDelegate
{

  private weak var parentAdapter: GADMAdapterAppLovin?

  @objc public init(parentAdapter: GADMAdapterAppLovin) {
    self.parentAdapter = parentAdapter
    super.init()
  }

  // MARK: - Ad Load Delegate
  public func adService(_ adService: ALAdService, didLoad ad: ALAd) {
    guard let parentAdapter else { return }
    GADMAdapterAppLovinUtils.log("Banner did load ad: \(ad)")
    parentAdapter.adView?.render(ad)
    parentAdapter.connector?.adapter(parentAdapter, didReceiveAdView: parentAdapter.adView)
  }

  public func adService(_ adService: ALAdService, didFailToLoadAdWithError code: Int32) {
    guard let parentAdapter else { return }
    let error = GADMAdapterAppLovinUtils.sdkError(withCode: Int(code))
    parentAdapter.connector?.adapter(parentAdapter, didFailAd: error)
  }

  // MARK: - Ad Display Delegate
  public func ad(_ ad: ALAd, wasDisplayedIn view: UIView) {
    GADMAdapterAppLovinUtils.log("Banner displayed")
  }

  public func ad(_ ad: ALAd, wasHiddenIn view: UIView) {
    GADMAdapterAppLovinUtils.log("Banner dismissed")
  }

  public func ad(_ ad: ALAd, wasClickedIn view: UIView) {
    guard let parentAdapter else { return }
    GADMAdapterAppLovinUtils.log("Banner clicked")
    parentAdapter.connector?.adapterDidGetAdClick(parentAdapter)
  }

  // MARK: - Ad View Event Delegate
  public func ad(_ ad: ALAd, didPresentFullscreenFor adView: ALAdView) {
    guard let parentAdapter else { return }
    GADMAdapterAppLovinUtils.log("Banner presented fullscreen")
    parentAdapter.connector?.adapterWillPresentFullScreenModal(parentAdapter)
  }

  public func ad(_ ad: ALAd, willDismissFullscreenFor adView: ALAdView) {
    guard let parentAdapter else { return }
    GADMAdapterAppLovinUtils.log("Banner will dismiss fullscreen")
    parentAdapter.connector?.adapterWillDismissFullScreenModal(parentAdapter)
  }

  public func ad(_ ad: ALAd, didDismissFullscreenFor adView: ALAdView) {
    guard let parentAdapter else { return }
    GADMAdapterAppLovinUtils.log("Banner did dismiss fullscreen")
    parentAdapter.connector?.adapterDidDismissFullScreenModal(parentAdapter)
  }

  public func ad(_ ad: ALAd, willLeaveApplicationFor adView: ALAdView) {
    guard let parentAdapter else { return }
    GADMAdapterAppLovinUtils.log("Banner left application")
    parentAdapter.connector?.adapterWillLeaveApplication(parentAdapter)
  }

  public func ad(
    _ ad: ALAd, didFailToDisplayIn adView: ALAdView, withError code: ALAdViewDisplayErrorCode
  ) {
    GADMAdapterAppLovinUtils.log("Banner failed to display: \(code.rawValue)")
  }
}
